cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java

```diff
@@ -0,0 +1,869 @@
+                                                    RebalanceOptions.RebalanceOptionsBuilder rebalanceOptionsBuilder) {
+        if (rebalanceAnnotation == RebalanceAnnotation.refresh) {
+            // the user fixed the error on the resource and want to "refresh", actually
+            // requesting a new rebalance proposal
+            return onNew(reconciliation, host, apiClient, rebalanceOptionsBuilder);
+        } else {
+            // stay in the current error state, actually failing the Future
+            return Future.failedFuture(clusterRebalance.getStatus().getConditions().get(0).getMessage());
+        }
+    }
+
+    /**
+     * This method handles the transition from {@code PendingProposal} state.
+     * It starts a periodic timer in order to check the status of the ongoing rebalance proposal processing on Cruise Control side.
+     * In order to do that, it calls the Cruise Control API for requesting the rebalance proposal.
+     * When the proposal is ready, the next state is {@code ProposalReady}.
+     * If the user sets the strimzi.io/rebalance=stop annotation, it stops to polling the Cruise Control API for requesting the rebalance proposal.
+     * If the user sets any other values for the strimzi.io/rebalance annotation, it is just ignored and the rebalance proposal request just continues.
+     * This method holds the lock until the rebalance proposal is ready or any exception is raised.
+     *
+     * @param reconciliation Reconciliation information
+     * @param host Cruise Control service to which sending the REST API requests
+     * @param apiClient Cruise Control REST API client instance
+     * @param clusterRebalance Current {@code KafkaRebalance} resource
+     * @param rebalanceAnnotation The current value for the strimzi.io/rebalance annotation
+     * @param rebalanceOptionsBuilder builder for the Cruise Control REST API client options
+     * @return a Future with the next {@code KafkaRebalanceStatus} bringing the state
+     */
+    private Future<KafkaRebalanceStatus> onPendingProposal(Reconciliation reconciliation,
+                                                           String host, CruiseControlApi apiClient,
+                                                           KafkaRebalance clusterRebalance,
+                                                           RebalanceAnnotation rebalanceAnnotation,
+                                                           RebalanceOptions.RebalanceOptionsBuilder rebalanceOptionsBuilder) {
+        Promise<KafkaRebalanceStatus> p = Promise.promise();
+        if (rebalanceAnnotation == RebalanceAnnotation.none) {
+            log.debug("{}: Arming Cruise Control rebalance proposal request timer", reconciliation);
+            vertx.setPeriodic(REBALANCE_POLLING_TIMER_MS, t -> {
+                resourceOperator.getAsync(clusterRebalance.getMetadata().getNamespace(), clusterRebalance.getMetadata().getName()).setHandler(getResult -> {
```

Comment:
This is fairly heavy use of the Kubernetes API if you poll this every few seconds. I wonder if this will cause any issues.

Reply:
It's a valid point. One thing we could do would be to utilise the watch we've already got to store the latest version when it changes and here to retrieve that. A little care other threading would be needed, but it wouldn't put any extra load on k8s.

Reply:
It could make sense but it means having a globally available `updatedRebalance` variable updated by the watcher every time it's fired inside a synchronized block and before acquiring the Vert.x lock for doing the reconcile. The above code could avoid calling K8S and using that global `updatedRebalance` in a sync block as well. My question is: if the watcher is raised more time, the global var will always contain the last updated one, but when the watcher code will be unlocked on the Vert.x lock for doing the reconcile, should it process the latest updated  or the one got when it was fired? I mean an old watch could operate on a more recent resource compared to when the watcher was raised.

Reply:
Anyway it needs more testing, I would leave it for later when we'll have more tests with upcoming PR.